[RISCV] Implement CodeGen Support for XCValu Extension in CV32E40P

realqhc opened the pull request:
Implement XCValu intrinsics and CodeGen for CV32E40P according to the specification.

This commit is part of a patch-set to upstream the vendor specific extensions of CV32E40P that need LLVM intrinsics to implement Clang builtins.

Contributors: @CharKeaney, @ChunyuLiao, @jeremybennett, @lewis-revill, @NandniJamnadas, @PaoloS02, @serkm, @simonpcook, @xingmingjie.

realqhc committed:
[RISCV] Implement Intrinsics and CodeGen Support for XCValu Extension in CV32E40P

Implement XCValu intrinsics for CV32E40P according to the
specification.

This commit is part of a patch-set to upstream the vendor specific
extensions of CV32E40P that need LLVM intrinsics to implement Clang
builtins.

Contributors: @CharKeaney, @ChunyuLiao, @jeremybennett, @lewis-revill,
@NandniJamnadas, @PaoloS02, @serkm, @simonpcook, @xingmingjie.

topperc commented on llvm/lib/Target/RISCV/RISCVExpandPseudoInsts.cpp:
```diff
@@ -547,6 +563,74 @@ bool RISCVPreRAExpandPseudo::expandLoadTLSGDAddress(
+  const MCInstrDesc &Desc = TII->get(Opcode);
+  BuildMI(MBB, MBBI, DL, Desc, DstReg)
+      .addReg(I)
+      .addImm(Log2_32_Ceil(J + 1) + 1);
```
Why do we need a pseudo instruction? Can't we use an SDNodeXForm to convert the immediate?

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCV.td:
```diff
@@ -662,6 +662,8 @@ let Predicates = [HasVendorXCVelw, IsRV32], hasSideEffects = 0,
 def cv_tuimm2 : TImmLeaf<XLenVT, [{return isUInt<2>(Imm);}]>;
 def cv_tuimm5 : TImmLeaf<XLenVT, [{return isUInt<5>(Imm);}]>;
 def cv_uimm10 : ImmLeaf<XLenVT, [{return isUInt<10>(Imm);}]>;
+def cv_uimm32: Operand<XLenVT>,
```
With the precedent set by `cv_uimm10`, `cv_tuimm5, etc., I would expect `cv_uimm32` to be isUint<32>(Imm), but its not.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCV.td:
```diff
@@ -704,3 +749,112 @@ let Predicates = [HasVendorXCVbitmanip, IsRV32] in {
+
+  def : Pat<(sra (add (XLenVT GPR:$rs1), (XLenVT GPR:$rs2)), uimm5:$imm5),
+            (CV_ADDN GPR:$rs1, GPR:$rs2, uimm5:$imm5)>;
+  def : Pat<(srl (add (XLenVT GPR:$rs1), (XLenVT GPR:$rs2)), uimm5:$imm5),
```
It's not clear to me from the spec that this is the right pattern. I would expect the addition and shift be evaluated with XLen+1 bits so the carry out from the addition is not lost before the shift.

The pattern you have here produces 0 for ((0x80000000+0x80000000)>>1) but I would expect the the cv.addun instruction would produce 0x80000000 when the same inputs and a shift of 1.

topperc replied:
Ok. I just expected it would work like the vaaddu instruction with RDN. The instruction seemed more useful with that behavior since it is hard to get the extract precision otherwise. But if that's not what the hardware does, then this pattern is correct.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCV.td:
```diff
@@ -704,3 +749,109 @@ let Predicates = [HasVendorXCVbitmanip, IsRV32] in {
+
+  def : Pat<(clip powerOf2Minus1:$upperBound, (XLenVT GPR:$rs1)),
+            (CV_CLIP GPR:$rs1, (trailing1sPlus1 imm:$upperBound))>;
+  def : Pat<(between (not GPR:$rs2), GPR:$rs2, (XLenVT GPR:$rs1)),
```
What if (not GPR:$rs2) is not less than GPR:$rs2? Then they aren't lower and upper bound.

Does the hardware implement the checks in this order
```
if rs1 <= -2^(Is2-1), rD = -2^(Is2-1),
else if rs1 >= 2^(Is2-1)-1, rD = 2^(Is2-1)-1,
else rD = rs1
```

If so then I think we can only match the pattern where the smax is done before the smin.

topperc replied:
alive2 proof that shows the order of smin/smax matter if you don't know that rs2 is positive. https://alive2.llvm.org/ce/z/dpHmEL

realqhc replied:
> @topperc Good point. The hardware actually treats GPR:$rs2 as unsigned, so it makes sense to generate this instruction only if rs2 is unsigned (e.g.: if it is an ABS node). The fix should require the node there to be unsigned.
```
cv.abs a0, a1
cv.clipr a0, a0, a1
```
Is this what we should generate to address this issue? If so, I have proposed a new pattern that generates this.

topperc replied:
By "unsigned" do you mean bit 31 must be zero?

PaoloS02 replied:
That is a fair point. I raised this issue with the hw group and we should have soon an update that clarifies this. I think that having bit 31 zero (so rs2 being positive) is a requirement for correctness of the operation. In that case we'd need a check on the operand value. I agree with separating this patch into two for now: intrinsics support and codegen.

realqhc committed:
address comments about negative input in cv_clipr

topperc commented on llvm/test/CodeGen/RISCV/xcvalu.ll:
```diff
@@ -0,0 +1,585 @@
+define i32 @clipr(i32 %a, i32 %b) {
+; CHECK-LABEL: clipr:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    cv.abs a0, a1
```
The isel you pattern you changed looks like it should absorb the abs, so why is it still here?

topperc commented on llvm/test/CodeGen/RISCV/xcvalu.ll:
```diff
@@ -0,0 +1,585 @@
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    cv.exthz a0, a0
+; CHECK-NEXT:    ret
+  %1 = call i32 @llvm.riscv.cv.alu.exthz(i32 %a)
```
Why do we need an intrinsic for exthz? Isn't this just AND with 0xffff?

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -1384,6 +1386,16 @@ RISCVTargetLowering::RISCVTargetLowering(const TargetMachine &TM,
     }
   }
 
+  if (Subtarget.hasVendorXCValu()) {
```
Can we pull the simple instructions out to a separate patch that can be easily approved?

realqhc committed:
Fix clipr and clipur according to the specification update.

https://github.com/pascalgouedo/cv32e40p/commit/5c9a14df76dc9087bd96f208821bdd349f4bd7e9

topperc commented:
Can we please split up this patch so I can approve the simple cases?

realqhc commented:
> Can we please split up this patch so I can approve the simple cases?

I have prepared the striped version in this pull request. https://github.com/llvm/llvm-project/pull/85603

realqhc committed:
remove the lshr pattern.
